---
title: "Assignment2"
author: "ChiaraOrlacchio"
date: "2024-03-01"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
suppressPackageStartupMessages({
  library(AnnotationHub) # to fetch genomes, annotations, etc.
  library(ensembldb)     # to manipulate ensembl gene annotations
  library(GenomicRanges) # data structure for genomic ranges 
  library(epiwraps)      # visualization
})
```

## 1. Using AnnotationHub, find and download the following annotations data:
```{r}
ah <- AnnotationHub()
````

#The Mouse EnsDb object
```{r}
q_MouseEnsDb <- query(ah, c("Mus Musculus", "EnsDb", "102", "GRCm38")) # to query the AH programmatically
q_MouseEnsDb
mouse_ensdb <- ah[["AH89211"]]
mouse_ensdb
```

#The mouse genome sequence in TwoBit
```{r}
q_MouseGenome <- query(ah, c("TwoBitFile", "Mus musculus", "GRCm38", "dna_sm"))
q_MouseGenome
mouse_genome <- ah[["AH49775"]]
mouse_genome
```

# Query AnnotationHub for Drosophila melanogaster genome sequence in TwoBit
```{r}
q_DrosophilaGenome <- query(ah, c("TwoBitFile", "Drosophila melanogaster", "BDGP6", "dna_sm"))
q_DrosophilaGenome
drosophila_genome <- ah[["AH49674"]]
drosophila_genome
```

## 2. Using the mouse EnsDb, find the following:

# How many different ensembl gene IDs and gene symbols are there for protein-coding genes?
```{r}
g <- genes(mouse_ensdb)
g
pcg <- g[which(g$gene_biotype=="protein_coding")] #protein coding genes

gene_ids <- unique(pcg$gene_id)
gene_symbol <- unique(pcg$symbol)

# Count the number of unique Ensembl gene IDs and gene symbols
num_ensembl_gene_ids <- length(gene_ids)
num_ensembl_gene_ids
num_gene_symbol <- length(gene_symbol)
num_gene_symbol
```
# Extract spliced lengths (exons) of protein-coding transcripts
```{r}
exsPerPct <- exonsBy(mouse_ensdb, column=c("tx_id","tx_biotype"), filter=TxBiotypeFilter("protein_coding"))
#exons length of each protein coding transcript
exons_length <- width(exsPerPct)

#sum of exons length of each pct
transcript_length <- sum(exons_length)
```

# Plot the distribution of spliced lengths
```{r}
hist(transcript_length, main = "Histogram of spliced Length", 
     xlab = "spliced Length", ylab = "Frequency", xlim = c(0, 15000), breaks = 300)
```















