---
title: "Assignment7"
author: "ChiaraOrlacchio"
date: "2024-04-19"
output: html_document
---

#In the same dataset of ATAC on chr19, plot 1) the insertion (i.e. ‘cuts’) profile of nucleosome-free fragments and 2) the centers of nucleosome-containing fragments, around the high-confidence motifs of two factors.

#You can choose your own factors of interest, or for instance use KLF4 and MAZ.

#Expected form of the answer: 2 figures (one for each factor), each containing the two signals around the motifs

#we should do NF and mono-nucleosome fragmnets and we should do 2 figures (one for each factor) with per figure two columnes (one for NF and one for the mono-NF)


```{r}
BiocManager::install("ETHZ-INS/epiwraps")
BiocManager::install("magick")
suppressPackageStartupMessages({
  library(epiwraps)
  library(AnnotationHub)
  library(MotifDb)
  library(universalmotif)
  library(ensembldb)
  library(ggplot2)
})

ah <- AnnotationHub(localHub=TRUE)
```


## Download the data

You can either start from the bam file:

```{r, eval=FALSE}
download.file("https://ethz-ins.org/content/w7/atac.chr19.bam", "atac.chr19.bam") #here only chr19
download.file("https://ethz-ins.org/content/w7/atac.chr19.bam.bai", "atac.chr19.bam.bai") 
```

## Preparing tracks

```{r}
bam <- "atac.chr19.bam"

# create a track using only nucleosome-free fragments, the number of cuts/insertion sites at each position
bam2bw(bam, output_bw = "NF_cuts.bw", paired=TRUE, binWidth=1L, type="ends", extend=2L, minFragLength=30, maxFragLength=120, shift=c(4L,-5L), forceSeqlevelsStyle = "Ensembl")


# create a track using only the (10bp) centers of mono-nucleosome fragments
bam2bw(bam, output_bw = "mono_centers.bw", paired=TRUE, binWidth=5L, minFragLength=140, shift=c(4L,-5L), maxFragLength=220, type="center", extent=10L, forceSeqlevelsStyle = "Ensembl")
```


## Obtaining the sites with a KLF4 motif

```{r}
# get KLF4 motif
motif_KLF4 <- MotifDb::query(MotifDb, c("KLF4","Mus"))[[1]]
motif2_KLF4 <- convert_motifs(motif_KLF4, class="TFBSTools-PFMatrix")
genome <- ah[["AH68356"]]
# get the sequence for chr19:
chr19 <- import(genome)["19"]

# find motif matches across chr19
moi_KLF4_ch19 <- motifmatchr::matchMotifs(motif2_KLF4, chr19, out="positions", p.cutoff=1e-5)[[1]]
# convert to GRanges
moi_KLF4_ch19 <- as(setNames(moi_KLF4_ch19,names(chr19)), "GRanges")
```

## Plotting signal around the KLF4 motif occurences

```{r, fig.width=8, fig.height=4}
# we prepare the list of tracks
tracks <- c("KLF4 - NF cuts"="NF_cuts.bw", "KLF4 - Mono centers"="mono_centers.bw")

# extract signals around the motif occurences
# we zoom in to 300bp around the motif centers, in windows of 5bp
sm_KLF4 <- signal2Matrix(tracks, moi_KLF4_ch19, w=5, extend=300)

# background normalization to make sure that we have comparable signals
# then we apply the normalization factors:
nf <- getNormFactors(tracks, useSeqLevels="19", nwind=5000L)
sm_KLF4 <- renormalizeSignalMatrices(sm_KLF4, scaleFactors = nf)
plotEnrichedHeatmaps(sm_KLF4, trim=0.95, colors = c("white","darkred"), minRowVal = 10, use_raster=FALSE) #higher density appeares in the heatmap where the cuts were made and for the upper graph one has to interpret the result that:For NF frafgments -> we cut around the TF bound to motif, therefor we see an increase where the cuts were made and a sink were the TF binds the motif (in the center); MN fragments -> we cunt around one Nucleosome therefor we see an increase where cuts were made and dump were the Nucleosome sits
```


## Obtaining the sites with a MAZ motif

```{r}
# get MAZ motif
motif_MAZ <- MotifDb::query(MotifDb, c("MAZ","Mus"))[[1]]
motif2_MAZ <- convert_motifs(motif_MAZ, class="TFBSTools-PFMatrix")
genome <- ah[["AH68356"]]
# get the sequence for chr19:
chr19 <- import(genome)["19"]

# find motif matches across chr19
moi_MAZ_chr19 <- motifmatchr::matchMotifs(motif2_MAZ, chr19, out="positions", p.cutoff=1e-5)[[1]]
# convert to GRanges
moi_MAZ_chr19 <- as(setNames(moi_MAZ_chr19,names(chr19)), "GRanges")
```

## Plotting signal around the motif occurences

```{r, fig.width=8, fig.height=4}
# we prepare the list of tracks
tracks <- c("MAZ - NF cuts"="NF_cuts.bw", "MAZ - Mono centers"="mono_centers.bw")

# extract signals around the motif occurences -> we zoom in to 300bp around the motif centers, in windows of 5bp
sm_MAZ <- signal2Matrix(tracks, moi_MAZ_chr19, w=5, extend=300)

# background normalization to make sure  we have comparable signals
# then we apply the normalization factors:
nf <- getNormFactors(tracks, useSeqLevels="19", nwind=5000L)
sm_MAZ <- renormalizeSignalMatrices(sm_MAZ, scaleFactors = nf)
plotEnrichedHeatmaps(sm_MAZ, trim=0.95, colors = c("white","darkred"), minRowVal = 10, use_raster=FALSE)
```


